compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -7014,26 +7014,48 @@ private BType checkInvocationArgs(BLangInvocation iExpr, List<BType> paramTypes,
     }
 
     private void checkArrayLibSortFuncArgs(BLangInvocation iExpr) {
-        if (iExpr.argExprs.size() <= 2 && !types.isOrderedType(iExpr.argExprs.get(0).getBType(), false)) {
-            dlog.error(iExpr.argExprs.get(0).pos, DiagnosticErrorCode.INVALID_SORT_ARRAY_MEMBER_TYPE,
-                       iExpr.argExprs.get(0).getBType());
+        List<BLangExpression> argExprs = iExpr.argExprs;
+
+        boolean hasKeyFunction = false;
+        BLangExpression keyFunction = null;
+
+        for (BLangExpression arg : argExprs) {
+            if (arg.getKind() == NodeKind.NAMED_ARGS_EXPR) {
+                BLangNamedArgsExpression argExpr = (BLangNamedArgsExpression) arg;
+                if (argExpr.name.value.equals("key")) {
+                    keyFunction = argExpr.expr;
+                    hasKeyFunction = true;
+                }
+            }
         }
 
-        if (iExpr.argExprs.size() != 3) {
+        if (!hasKeyFunction && argExprs.size() == 3) {
+            hasKeyFunction = true;
+        }
+
+        BLangExpression arrExpr = argExprs.get(0);
+        BType arrType = arrExpr.getBType();
+        boolean isOrderedType = types.isOrderedType(arrType, false);
+        if (!hasKeyFunction) {
+            if (!isOrderedType) {
+                dlog.error(arrExpr.pos, DiagnosticErrorCode.INVALID_SORT_ARRAY_MEMBER_TYPE, arrType);
+            }
             return;
         }
 
-        BLangExpression keyFunction = iExpr.argExprs.get(2);
+        if (keyFunction == null) {
+            keyFunction = iExpr.argExprs.get(2);
+        }
+
```

Comment:
We can remove the local variable `hasKeyFunction` by changing the logic like this.
```ballerina
        BLangExpression keyFunction = null;

        for (int i = 0; i < argExprs.size(); i++) {
            BLangExpression arg = argExprs.get(i);
            if (arg.getKind() == NodeKind.NAMED_ARGS_EXPR) {
                BLangNamedArgsExpression argExpr = (BLangNamedArgsExpression) arg;
                if (argExpr.name.value.equals("key")) {
                    keyFunction = argExpr.expr;
                    break;
                }
            } else if (i == 2) { // last argument, `2` can be defined as a constant
                keyFunction = arg;
                break;
            }
        }

        BLangExpression arrExpr = argExprs.get(0);
        BType arrType = arrExpr.getBType();
        boolean isOrderedType = types.isOrderedType(arrType, false);
        if (keyFunction == null) {
            if (!isOrderedType) {
                dlog.error(arrExpr.pos, DiagnosticErrorCode.INVALID_SORT_ARRAY_MEMBER_TYPE, arrType);
            }
            return;
        }

        BType keyFunctionType = keyFunction.getBType();
```